hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ScmClient.java

```diff
@@ -111,7 +112,14 @@ public Map<Long, Pipeline> getContainerLocations(Iterable<Long> containerIds,
       containerLocationCache.invalidateAll(containerIds);
     }
     try {
-      return containerLocationCache.getAll(containerIds);
+      Map<Long, Pipeline> result = containerLocationCache.getAll(containerIds);
+      // Don't keep empty pipelines in the cache.
+      List<Long> emptyPipelines = result.entrySet().stream()
```

Comment:
Will we land up scanning all containers for each get request?

Reply:
Only the containers for this particular request. (Each request can bind to more than one container if a key has 2 or more blocks on 2 or more containers).

Reply:
As fas as I know, this method is only called in the `KeyManagerImpl` and `Iterable<Long> containerIds` parameter contains only the containerIds for a particular `OmKeyInfo`.

Across how many containers can a key span? What's the worst case scenario?

Reply:
It's simply linear to the number of blocks the key has. E.g. a key with 10 blocks can span 10 containers. Anyway, the cost is contained by the number of blocks, just like before.

Reply:
What about a background thread that checks and removes empty pipelines from cache, every 10 or 20 mins? After looking at `KeyManagerImpl#refreshPipelineFromCache()`, it's getting the containerIds for a list of Keys. The number of container pipelines can be too much to iterate on every request.

Reply:
@xBis7 we can add a background refresh loop but I am not sure it will lead to significant improvement, would something interesting to test and implement if we see a difference.